src/kernels/kernelProvider.base.ts

```diff
@@ -8,38 +8,58 @@ import { IVSCodeNotebook } from '../platform/common/application/types';
 import { traceInfoIfCI, traceVerbose, traceWarning } from '../platform/logging';
 import { getDisplayPath } from '../platform/common/platform/fs-paths';
 import { IAsyncDisposable, IAsyncDisposableRegistry, IDisposableRegistry } from '../platform/common/types';
-import { noop } from '../platform/common/utils/misc';
-import { IKernel, IKernelProvider, KernelOptions } from './types';
+import { isUri, noop } from '../platform/common/utils/misc';
+import { IKernel, IKernelProvider, INotebookKernel, KernelOptions } from './types';
 
 export abstract class BaseKernelProvider implements IKernelProvider {
     /**
      * Use a separate dictionary to track kernels by Notebook, so that
      * the ref to kernel is lost when the notebook is closed.
      */
-    private readonly kernelsByNotebook = new WeakMap<NotebookDocument, { options: KernelOptions; kernel: IKernel }>();
+    protected readonly kernelsByNotebook = new WeakMap<
+        NotebookDocument,
+        { options: KernelOptions; kernel: INotebookKernel }
+    >();
     /**
      * The life time of kernels not tied to a notebook will be managed by callers of the API.
      * Where as if a kernel is tied to a notebook, then the kernel dies along with notebooks.
      */
     private readonly kernelsByUri = new Map<string, { options: KernelOptions; kernel: IKernel }>();
     private readonly pendingDisposables = new Set<IAsyncDisposable>();
     protected readonly _onDidRestartKernel = new EventEmitter<IKernel>();
+    protected readonly _onDidRestartNotebookKernel = new EventEmitter<INotebookKernel>();
     protected readonly _onDidStartKernel = new EventEmitter<IKernel>();
+    protected readonly _onDidStartNotebookKernel = new EventEmitter<INotebookKernel>();
     protected readonly _onDidCreateKernel = new EventEmitter<IKernel>();
+    protected readonly _onDidCreateNotebookKernel = new EventEmitter<INotebookKernel>();
```

Comment:
Do we really need these duplicate events
Can we not have the same event. & when using this we can just check if `kernel.notebook` exists, and if it does then typescript will automatically treat the type as INotebookKernel (after all INotebookKernel extends IKernel).

Else feels like a lot of duplication.

Reply:
This is currently intentional as we are falling into the same problem as before that the core system knows the existence of `IKernel` created by Data Wrangler. Ideally the core system only needs to care about events for `INotebookKernel` and in very few cases that we need to handle both `INotebookKernel` and `IKernel` created by 3rd party.